src/azul/reindexer.py

```diff
@@ -0,0 +1,137 @@
+"""
+
+from collections import defaultdict
+from concurrent.futures import Future, ThreadPoolExecutor
+from functools import partial
+import json
+import logging
+from pprint import PrettyPrinter
+from urllib.error import HTTPError
+from urllib.parse import urlparse, urlencode, parse_qs
+from urllib.request import Request, urlopen
+from uuid import uuid4
+
+from azul import config
+
+logger = logging.getLogger(__name__)
+
+plugin = config.plugin()
+
+
+class Defaults:
+    dss_url = config.dss_endpoint
+    indexer_url = "https://" + config.api_lambda_domain('indexer') + "/"
+    es_query = plugin.dss_subscription_query
+    num_workers = 16
+
+
+def post_bundle(bundle_fqid, es_query, indexer_url):
+    """
+    Send a mock DSS notification to the indexer
+    """
+    bundle_uuid, _, bundle_version = bundle_fqid.partition('.')
+    simulated_event = {
+        "query": es_query,
+        "subscription_id": str(uuid4()),
+        "transaction_id": str(uuid4()),
+        "match": {
+            "bundle_uuid": bundle_uuid,
+            "bundle_version": bundle_version
+        }
+    }
+    body = json.dumps(simulated_event).encode('utf-8')
+    request = Request(indexer_url, body)
+    request.add_header("content-type", "application/json")
+    with urlopen(request) as f:
+        return f.read()
+
+
+class Reindexer(object):
+    def __init__(self, indexer_url: str = Defaults.indexer_url,
+                 dss_url: str = Defaults.dss_url,
+                 es_query: dict = Defaults.es_query,
+                 will_sync_notifications: bool = False,
+                 number_of_workers: int = Defaults.num_workers,
+                 test_name: str = None,
+                 is_dry_run: bool = None):
+
+        self.number_of_workers = number_of_workers
+        self.will_sync_notifications = will_sync_notifications
+        self.es_query = es_query
+        self.dss_url = dss_url
+        self.indexer_url = indexer_url
+        self.test_name = test_name
+        self.is_dry_run = is_dry_run
+
+    def reindex(self):
+        errors = defaultdict(int)
+        missing = {}
+        indexed = 0
+        total = 0
+
+        logger.info('Querying DSS using %s', json.dumps(self.es_query, indent=4))
+        dss_client = config.dss_client(dss_endpoint=self.dss_url)
+        # noinspection PyUnresolvedReferences
+        response = dss_client.post_search.iterate(es_query=self.es_query, replica="aws")
+        bundle_fqids = [r['bundle_fqid'] for r in response]
+        logger.info("Bundle FQIDs to index: %i", len(bundle_fqids))
+
+        with ThreadPoolExecutor(max_workers=self.number_of_workers, thread_name_prefix='pool') as tpe:
+
+            def attempt(bundle_fqid, i):
+                try:
+                    logger.info("Bundle %s, attempt %i: Sending notification", bundle_fqid, i)
+                    url = urlparse(self.indexer_url)
+                    if self.will_sync_notifications is not None:
```

Comment:
After our discussion about this comment, Hannes decided to put this change on hold. All indexer parameters should be passed the same way, but there's a discrepancy on how `sync` and `test_name` are passed. `Sync` is passed by a url param and `test_name` is passed thru the notification response. However,  both parameters can't be passed by notification, since the notification must be read at the `/` route method (which we don't want). And passing both parameters by url is impossible, since the `test_name` will be lost in the queue.